storage/distributedstorage.py

```diff
@@ -28,15 +28,27 @@ def wrapper(self, locations, *args, **kwargs):
     return wrapper
 
 
-def _call_endtoend(unbound_func, requires_write=False):
+def _call_endtoend(unbound_func, requires_write=False, requires_all_available=False):
+    """Wrapper to call a storage function for all locations.
+    Should raise exception if requires_all_available is set and one of the calls fails.
+    None otherwise.
+    """
+
     @wraps(unbound_func)
     def wrapper(self, locations, *args, **kwargs):
         if requires_write:
             assert not self.readonly_mode
 
-        for storage in self._storages:
-            storage_func = getattr(storage, unbound_func.__name__)
-            storage_func(*args, **kwargs)
+        for location in locations:
+            storage = self._storages[location]
+            try:
+                storage_func = getattr(storage, unbound_func.__name__)
+                storage_func(*args, **kwargs)
+            except Exception as e:
+                if requires_all_available or len(locations) == 1:
```

Comment:
If `requires_all_available` is set and at least one is still available, then the health check should pass. AFAIK that was the ask for PROJQUAY-5074

Reply:
I was more thinking about the case where `requires_all_available` is unset and you have two or three different locations and each one raises and exception, wouldn't it still pass because each time an exception is thrown `len(locations) != 1`?